// react-native-sensitive-info/lib/commonjs/internal/options.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_SERVICE = exports.DEFAULT_ACCESS_CONTROL = void 0;
exports.normalizeOptions = normalizeOptions;
exports.normalizePromptedReadOptions = normalizePromptedReadOptions;
exports.normalizeStorageScopeOptions = normalizeStorageScopeOptions;
const DEFAULT_SERVICE = exports.DEFAULT_SERVICE = 'default';
const DEFAULT_ACCESS_CONTROL = exports.DEFAULT_ACCESS_CONTROL = 'secureEnclaveBiometry';

/**
 * Normalises user supplied options by applying defaults and pruning `undefined` fields.
 *
 * Keeping the shape deterministic reduces marshalling work on the native side and makes
 * memoisation on hot paths cheap (stable JSON fingerprints).
 */
function normalizeOptions(options) {
  if (options == null) {
    return {
      service: DEFAULT_SERVICE,
      accessControl: DEFAULT_ACCESS_CONTROL
    };
  }
  const normalized = {
    service: options.service ?? DEFAULT_SERVICE,
    accessControl: options.accessControl ?? DEFAULT_ACCESS_CONTROL
  };
  if (options.iosSynchronizable !== undefined) {
    normalized.iosSynchronizable = options.iosSynchronizable;
  }
  if (options.keychainGroup !== undefined) {
    normalized.keychainGroup = options.keychainGroup;
  }
  if (options.authenticationPrompt !== undefined) {
    normalized.authenticationPrompt = options.authenticationPrompt;
  }
  return normalized;
}
function normalizeStorageScopeOptions(options) {
  const normalized = {
    service: options?.service ?? DEFAULT_SERVICE
  };
  if (options?.iosSynchronizable !== undefined) {
    normalized.iosSynchronizable = options.iosSynchronizable;
  }
  if (options?.keychainGroup !== undefined) {
    normalized.keychainGroup = options.keychainGroup;
  }
  return normalized;
}
function normalizePromptedReadOptions(options) {
  const normalized = normalizeStorageScopeOptions(options);
  if (options?.authenticationPrompt !== undefined) {
    normalized.authenticationPrompt = options.authenticationPrompt;
  }
  return normalized;
}
