// react-native-sensitive-info/lib/commonjs/core/access-control.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.canUseAccessControl = canUseAccessControl;
exports.canUseAccessControlSync = canUseAccessControlSync;
var _storage = require("./storage.js");
/**
 * Pure mapping table: which capability does each {@link AccessControl} policy require?
 *
 * Kept as a `const` literal lookup (not a `switch`, not a `Map`) so the minifier can constant-fold
 * it and dead-code-eliminate unused branches. Internal helper — not exported from the public API.
 *
 * @internal
 */
const POLICY_PREDICATES = {
  secureEnclaveBiometry: l => (l.secureEnclave || l.strongBox) && l.biometryStatus === 'available',
  biometryCurrentSet: l => l.biometryStatus === 'available',
  biometryAny: l => l.biometryStatus === 'available',
  devicePasscode: l => l.deviceCredential,
  none: () => true
};

/**
 * Synchronous variant of {@link canUseAccessControl} for callers that already hold a
 * {@link SecurityAvailability} snapshot (e.g. inside a render that consumes
 * {@link useSecurityAvailability}).
 *
 * Pure function — no native call, no IPC round-trip. Safe to call inside React render paths.
 *
 * @param policy - The {@link AccessControl} policy you intend to write with.
 * @param levels - Capability snapshot, typically from {@link getSupportedSecurityLevels} or
 *   {@link useSecurityAvailability}.
 * @returns `true` when the device currently satisfies the policy's requirements; `false` when a
 *   write would fail or be silently downgraded.
 *
 * @example
 * ```tsx
 * const { data: caps } = useSecurityAvailability()
 * const canEnable = caps ? canUseAccessControlSync('secureEnclaveBiometry', caps) : false
 * return <Toggle disabled={!canEnable} />
 * ```
 *
 * @see {@link canUseAccessControl}
 * @public
 */
function canUseAccessControlSync(policy, levels) {
  return POLICY_PREDICATES[policy](levels);
}

/**
 * Predicts whether a given {@link AccessControl} policy can be satisfied on the current device
 * **right now** — useful for gating biometric toggles before attempting a write that would
 * otherwise fail or be silently downgraded.
 *
 * Internally maps the requested policy onto the {@link SecurityAvailability} snapshot returned by
 * {@link getSupportedSecurityLevels}. Pass `levels` to skip the native round-trip.
 *
 * @param policy - The {@link AccessControl} policy you intend to write with.
 * @param levels - Optional pre-fetched capability snapshot. When omitted, the helper calls
 *   {@link getSupportedSecurityLevels} internally.
 * @returns Resolves to `true` when the policy can be applied, `false` otherwise. Resolves to
 *   `false` (rather than throwing) for `'unknown'` biometry status — gate UI off availability
 *   instead.
 *
 * @remarks
 * This is a *predictive* check, not a guarantee — the user could change biometric settings
 * between this call and the subsequent write. Always handle {@link KeyInvalidatedError} and
 * {@link AuthenticationCanceledError} on the write path as well.
 *
 * @example
 * ```ts
 * if (await canUseAccessControl('secureEnclaveBiometry')) {
 *   await setItem('session', token, { accessControl: 'secureEnclaveBiometry' })
 * } else {
 *   await setItem('session', token, { accessControl: 'devicePasscode' })
 * }
 * ```
 *
 * @see {@link canUseAccessControlSync}
 * @see {@link getSupportedSecurityLevels}
 * @public
 */
async function canUseAccessControl(policy, levels) {
  const snapshot = levels ?? (await (0, _storage.getSupportedSecurityLevels)());
  return canUseAccessControlSync(policy, snapshot);
}
