// react-native-sensitive-info/lib/commonjs/core/storage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SensitiveInfo = void 0;
exports.clearService = clearService;
exports.deleteItem = deleteItem;
exports.getAllItems = getAllItems;
exports.getItem = getItem;
exports.getKeyVersion = getKeyVersion;
exports.getSupportedSecurityLevels = getSupportedSecurityLevels;
exports.hasItem = hasItem;
exports.rotateKeys = rotateKeys;
exports.setItem = setItem;
var _errors = require("../errors.js");
var _native = _interopRequireDefault(require("../internal/native.js"));
var _options = require("../internal/options.js");
var _validate = require("../internal/validate.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Strongly typed façade around the underlying Nitro native object.
 *
 * Each member mirrors a top-level export: prefer the named exports for tree-shaking, or use this
 * interface when injecting/mocking the storage surface (e.g. in tests).
 */

/**
 * Encrypts and persists a secret value in the platform secure storage.
 *
 * @param key   - Stable identifier for the entry (non-empty). Re-using a key overwrites the prior value.
 * @param value - UTF-8 string to encrypt. Encode binary data as base64 before passing.
 * @param options - Storage policy. See {@link SensitiveInfoOptions} for defaults.
 * @returns The {@link MutationResult} containing the metadata that was actually applied (the
 * platform may downgrade the requested `accessControl`).
 *
 * @throws {@link AuthenticationCanceledError} if the user dismisses the biometric prompt.
 * @throws {@link KeyInvalidatedError} if the device's biometric set changed since the last write.
 * @throws {@link SensitiveInfoError} for any other native failure.
 *
 * @example
 * ```ts
 * await setItem('session-token', token, {
 *   service: 'com.example.auth',
 *   accessControl: 'secureEnclaveBiometry',
 *   authenticationPrompt: { title: 'Save your session' },
 * })
 * ```
 *
 * @see {@link SensitiveInfoSetRequest}
 * @see {@link MutationResult}
 */
async function setItem(key, value, options) {
  (0, _validate.validateKey)(key);
  (0, _validate.validateValue)(value);
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const payload = {
    key,
    value,
    ...(0, _options.normalizeOptions)(options)
  };
  try {
    return await native.setItem(payload);
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Retrieves a previously stored secret.
 *
 * @param key - Identifier of the entry to fetch.
 * @param options - Storage scoping plus an optional `includeValue` flag (defaults to `true`).
 *   Pass `includeValue: false` to fetch metadata only — cheaper, and on iOS this avoids the
 *   biometric prompt on auth-gated entries.
 * @returns The {@link SensitiveInfoItem}, or `null` when no entry exists for `key`.
 *
 * @remarks
 * `NotFoundError` is intentionally swallowed and surfaced as `null`. Every other native error is
 * mapped to its typed {@link SensitiveInfoError} subclass and re-thrown.
 *
 * @throws {@link AuthenticationCanceledError} if the user dismisses the biometric prompt.
 * @throws {@link IntegrityViolationError} if the on-disk record fails its HMAC check.
 * @throws {@link KeyInvalidatedError} if the entry's biometric guard is no longer satisfiable.
 *
 * @example
 * ```ts
 * const item = await getItem('session-token', { service: 'com.example.auth' })
 * if (item) console.log(item.value, item.metadata.securityLevel)
 * ```
 *
 * @see {@link SensitiveInfoGetRequest}
 * @see {@link SensitiveInfoItem}
 */
async function getItem(key, options) {
  (0, _validate.validateKey)(key);
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const includeValue = options?.includeValue ?? true;
  const payload = {
    key,
    includeValue,
    ...(includeValue ? (0, _options.normalizePromptedReadOptions)(options) : (0, _options.normalizeStorageScopeOptions)(options))
  };
  try {
    return await native.getItem(payload);
  } catch (error) {
    if ((0, _errors.isNotFoundError)(error)) return null;
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Cheap existence check that never decrypts the value.
 *
 * @param key - Identifier to look up.
 * @param options - Storage scoping. `accessControl` / `authenticationPrompt` are ignored here —
 *   `hasItem` is designed to stay silent, even for biometric-protected entries.
 * @returns `true` when an entry exists for the key, `false` otherwise.
 *
 * @throws {@link SensitiveInfoError} for unexpected native failures (storage IO, etc.).
 *
 * @example
 * ```ts
 * if (await hasItem('session-token', { service: 'com.example.auth' })) {
 *   // skip onboarding
 * }
 * ```
 *
 * @see {@link SensitiveInfoHasRequest}
 */
async function hasItem(key, options) {
  (0, _validate.validateKey)(key);
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const payload = {
    key,
    ...(0, _options.normalizeStorageScopeOptions)(options)
  };
  try {
    return await native.hasItem(payload);
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Deletes a stored secret. Idempotent — calling on a missing key resolves with `false` rather
 * than throwing.
 *
 * @param key - Identifier of the entry to delete.
 * @param options - Storage scoping. The `accessControl`/`authenticationPrompt` from the original
 *   write are not required for deletion.
 * @returns `true` when an entry was removed, `false` when the key did not exist.
 *
 * @throws {@link SensitiveInfoError} for unexpected native failures.
 *
 * @example
 * ```ts
 * await deleteItem('session-token', { service: 'com.example.auth' })
 * ```
 *
 * @see {@link SensitiveInfoDeleteRequest}
 */
async function deleteItem(key, options) {
  (0, _validate.validateKey)(key);
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const payload = {
    key,
    ...(0, _options.normalizeStorageScopeOptions)(options)
  };
  try {
    return await native.deleteItem(payload);
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Enumerates every entry stored under the configured service namespace.
 *
 * @param options - Pass `{ includeValues: true }` to decrypt and return values; defaults to
 *   metadata-only for performance and to avoid biometric prompts on protected entries. Prompt
 *   strings are only forwarded when values are requested.
 * @returns Array of {@link SensitiveInfoItem}. Returns `[]` when the service is empty.
 *
 * @throws {@link AuthenticationCanceledError} when `includeValues: true` and the user cancels.
 * @throws {@link SensitiveInfoError} for unexpected native failures.
 *
 * @example
 * ```ts
 * const items = await getAllItems({ service: 'com.example.auth' })
 * console.log(items.map((i) => i.key))
 * ```
 *
 * @see {@link SensitiveInfoEnumerateRequest}
 */
async function getAllItems(options) {
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const payload = {
    includeValues: options?.includeValues ?? false,
    ...(options?.includeValues === true ? (0, _options.normalizePromptedReadOptions)(options) : (0, _options.normalizeStorageScopeOptions)(options))
  };
  try {
    return await native.getAllItems(payload);
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Removes every entry associated with a service.
 *
 * @param options - Storage scoping (only `service`/`keychainGroup`/`iosSynchronizable` matter).
 * @returns Resolves when the service has been emptied.
 *
 * @remarks This is **non-recoverable** — there is no undo. Prefer this over deleting keys one by
 * one when wiping a feature on logout.
 *
 * @throws {@link SensitiveInfoError} for unexpected native failures.
 *
 * @example
 * ```ts
 * await clearService({ service: 'com.example.auth' })
 * ```
 */
async function clearService(options) {
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  try {
    return await native.clearService((0, _options.normalizeStorageScopeOptions)(options));
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Inspects which security primitives are available on the current device.
 *
 * Use this to drive feature flags (e.g. show "Enable Face ID" only when `biometry === true`)
 * before writing entries that require those capabilities.
 *
 * @returns A {@link SecurityAvailability} snapshot — `secureEnclave` is iOS-only, `strongBox` is
 * Android-only.
 *
 * @throws {@link SensitiveInfoError} for unexpected native failures.
 *
 * @example
 * ```ts
 * const caps = await getSupportedSecurityLevels()
 * if (caps.biometry) enableBiometricUnlock()
 * ```
 *
 * @see {@link useSecurityAvailability} for the React-hook variant.
 */
async function getSupportedSecurityLevels() {
  const native = (0, _native.default)();
  try {
    return await native.getSupportedSecurityLevels();
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Rotates the master key for the given service.
 *
 * @param options - Pass `{ reEncryptEagerly: true }` to re-encrypt every existing entry in the
 *   same call; defaults to lazy rotation (entries are migrated on next read/write).
 * @returns A {@link RotationResult} with `previousVersion`, `newVersion`, and `reEncryptedCount`.
 *
 * @remarks Eager rotation may trigger one biometric prompt **per protected entry** — prefer the
 * default lazy mode unless you need to migrate ciphertext immediately for compliance reasons.
 *
 * @throws {@link RotationFailedError} when the native layer cannot generate a new key.
 * @throws {@link AuthenticationCanceledError} during eager rotation when the user cancels.
 *
 * @example
 * ```ts
 * const { newVersion } = await rotateKeys({ service: 'com.example.auth' })
 * console.log(`rotated to v${newVersion}`)
 * ```
 *
 * @see {@link RotateKeysRequest}
 * @see {@link useKeyRotation}
 */
async function rotateKeys(options) {
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  const payload = {
    reEncryptEagerly: options?.reEncryptEagerly ?? false,
    ...(options?.reEncryptEagerly === true ? (0, _options.normalizePromptedReadOptions)(options) : (0, _options.normalizeStorageScopeOptions)(options))
  };
  try {
    return await native.rotateKeys(payload);
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Returns the currently active key version for the given service.
 *
 * @param options - Storage scoping. `authenticationPrompt` is ignored here — this call should
 *   never trigger biometrics.
 * @returns A non-negative integer. `0` indicates a legacy entry that has not been rotated yet.
 *
 * @throws {@link SensitiveInfoError} for unexpected native failures.
 *
 * @example
 * ```ts
 * const version = await getKeyVersion({ service: 'com.example.auth' })
 * if (version === 0) await rotateKeys({ service: 'com.example.auth' })
 * ```
 *
 * @see {@link rotateKeys}
 */
async function getKeyVersion(options) {
  (0, _validate.validateService)(options);
  const native = (0, _native.default)();
  try {
    return await native.getKeyVersion((0, _options.normalizeStorageScopeOptions)(options));
  } catch (error) {
    throw (0, _errors.toSensitiveInfoError)(error);
  }
}

/**
 * Convenient namespace exposing the secure storage surface. Named exports are the preferred way
 * to consume the library for optimal tree-shaking.
 */
const SensitiveInfo = exports.SensitiveInfo = {
  setItem,
  getItem,
  hasItem,
  deleteItem,
  getAllItems,
  clearService,
  getSupportedSecurityLevels,
  rotateKeys,
  getKeyVersion
};
