// react-native-sensitive-info/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AuthenticationCanceledError", {
  enumerable: true,
  get: function () {
    return _errors.AuthenticationCanceledError;
  }
});
Object.defineProperty(exports, "ErrorCode", {
  enumerable: true,
  get: function () {
    return _errors.ErrorCode;
  }
});
Object.defineProperty(exports, "IntegrityViolationError", {
  enumerable: true,
  get: function () {
    return _errors.IntegrityViolationError;
  }
});
Object.defineProperty(exports, "InvalidArgumentError", {
  enumerable: true,
  get: function () {
    return _errors.InvalidArgumentError;
  }
});
Object.defineProperty(exports, "KeyInvalidatedError", {
  enumerable: true,
  get: function () {
    return _errors.KeyInvalidatedError;
  }
});
Object.defineProperty(exports, "NotFoundError", {
  enumerable: true,
  get: function () {
    return _errors.NotFoundError;
  }
});
Object.defineProperty(exports, "RotationFailedError", {
  enumerable: true,
  get: function () {
    return _errors.RotationFailedError;
  }
});
Object.defineProperty(exports, "SensitiveInfo", {
  enumerable: true,
  get: function () {
    return _storage.SensitiveInfo;
  }
});
Object.defineProperty(exports, "SensitiveInfoError", {
  enumerable: true,
  get: function () {
    return _errors.SensitiveInfoError;
  }
});
Object.defineProperty(exports, "canUseAccessControl", {
  enumerable: true,
  get: function () {
    return _accessControl.canUseAccessControl;
  }
});
Object.defineProperty(exports, "canUseAccessControlSync", {
  enumerable: true,
  get: function () {
    return _accessControl.canUseAccessControlSync;
  }
});
Object.defineProperty(exports, "clearService", {
  enumerable: true,
  get: function () {
    return _storage.clearService;
  }
});
Object.defineProperty(exports, "deleteItem", {
  enumerable: true,
  get: function () {
    return _storage.deleteItem;
  }
});
Object.defineProperty(exports, "getAllItems", {
  enumerable: true,
  get: function () {
    return _storage.getAllItems;
  }
});
Object.defineProperty(exports, "getItem", {
  enumerable: true,
  get: function () {
    return _storage.getItem;
  }
});
Object.defineProperty(exports, "getKeyVersion", {
  enumerable: true,
  get: function () {
    return _storage.getKeyVersion;
  }
});
Object.defineProperty(exports, "getSupportedSecurityLevels", {
  enumerable: true,
  get: function () {
    return _storage.getSupportedSecurityLevels;
  }
});
Object.defineProperty(exports, "hasItem", {
  enumerable: true,
  get: function () {
    return _storage.hasItem;
  }
});
Object.defineProperty(exports, "isAuthenticationCanceledError", {
  enumerable: true,
  get: function () {
    return _errors.isAuthenticationCanceledError;
  }
});
Object.defineProperty(exports, "isIntegrityViolationError", {
  enumerable: true,
  get: function () {
    return _errors.isIntegrityViolationError;
  }
});
Object.defineProperty(exports, "isInvalidArgumentError", {
  enumerable: true,
  get: function () {
    return _errors.isInvalidArgumentError;
  }
});
Object.defineProperty(exports, "isKeyInvalidatedError", {
  enumerable: true,
  get: function () {
    return _errors.isKeyInvalidatedError;
  }
});
Object.defineProperty(exports, "isNotFoundError", {
  enumerable: true,
  get: function () {
    return _errors.isNotFoundError;
  }
});
Object.defineProperty(exports, "isRotationFailedError", {
  enumerable: true,
  get: function () {
    return _errors.isRotationFailedError;
  }
});
Object.defineProperty(exports, "rotateKeys", {
  enumerable: true,
  get: function () {
    return _storage.rotateKeys;
  }
});
Object.defineProperty(exports, "setItem", {
  enumerable: true,
  get: function () {
    return _storage.setItem;
  }
});
var _accessControl = require("./core/access-control.js");
var _storage = require("./core/storage.js");
var _errors = require("./errors.js");
