// react-native-sensitive-info/lib/module/hooks/types.js
"use strict";

/**
 * Options accepted by the {@link HookError} constructor.
 */

/**
 * Error wrapper surfaced by the public hooks.
 *
 * Carries the original native error in `cause` plus extra UI-friendly context (`operation`,
 * `hint`) so consumers can build rich error states without unwrapping the underlying
 * {@link SensitiveInfoError}.
 *
 * @example
 * ```tsx
 * const { error } = useSecret('token', { service: 'auth' })
 * if (error) console.warn(`[${error.operation}] ${error.message} — ${error.hint}`)
 * ```
 */
// biome-ignore lint/suspicious/noUnsafeDeclarationMerging: intentional type-only `cause` augmentation via interface merging; see the `HookError` interface declaration below.
export class HookError extends Error {
  /** Identifier of the hook operation that failed (e.g. `useSecret.save`). */

  /** UI-facing remediation hint (e.g. `'Ask the user to retry biometrics.'`). */

  /**
   * @param message - Human-readable description of the failure.
   * @param options - Additional metadata; see {@link HookErrorOptions}.
   */
  constructor(message, options = {}) {
    super(message);
    this.name = 'HookError';
    this.operation = options.operation;
    this.hint = options.hint;
    // Define `cause` manually instead of passing it to `super()` so this
    // compiles cleanly under TS configs whose `lib` predates ES2022 (where
    // the second `Error` constructor argument was introduced), while keeping
    // the property non-enumerable to match the native ES2022 `Error` constructor.
    //
    // We deliberately use `'cause' in options` (rather than a value check)
    // so that `new HookError(msg, { cause: undefined })` still installs the
    // property — matching native `new Error(msg, { cause: undefined })`.
    if ('cause' in options) {
      Object.defineProperty(this, 'cause', {
        value: options.cause,
        writable: true,
        configurable: true,
        enumerable: false
      });
    }
  }
}

/**
 * Type-only declaration merge that exposes the optional `cause` property on
 * {@link HookError} for consumers compiling with a `tsconfig` `lib` target
 * that predates ES2022 (where `Error.cause` was added). Using interface
 * merging—rather than a `declare` class field—keeps the source fully erasable
 * by Babel/SWC and avoids emitting an enumerable own property.
 */

/**
 * Canonical async state contract returned by most hooks.
 *
 * @typeParam T - Shape of the resolved value (e.g. `string`, {@link SensitiveInfoItem}).
 */

/**
 * Async state contract used by operations that do not emit data (e.g. `clearService`).
 */

/**
 * Successful outcome of a hook mutation helper. Use the `success` discriminant to narrow.
 */

/**
 * Failure outcome of a hook mutation helper. Use the `success` discriminant to narrow.
 */

/**
 * Discriminated union returned by hook mutation helpers (`saveSecret`, `clearAll`, …).
 *
 * Mutations never throw — they always resolve with this union so component code can branch with
 * a simple `if (!result.success)` check.
 *
 * @example
 * ```tsx
 * const result = await saveSecret(value)
 * if (!result.success) showToast(result.error.message)
 * ```
 */

/**
 * Factory used to initialise {@link AsyncState} values.
 *
 * @typeParam T - Shape of the eventual data payload.
 * @returns A frozen initial state with `isLoading: true` and no data/error.
 *
 * @example
 * ```ts
 * const [state, setState] = useState(() => createInitialAsyncState<string>())
 * ```
 */
export function createInitialAsyncState() {
  return {
    data: null,
    error: null,
    isLoading: true,
    isPending: false
  };
}

/**
 * Factory used to initialise {@link VoidAsyncState} values.
 *
 * @returns Initial state with `isLoading: false` and no error — mutations idle by default.
 *
 * @example
 * ```ts
 * const [state, setState] = useState(() => createInitialVoidState())
 * ```
 */
export function createInitialVoidState() {
  return {
    error: null,
    isLoading: false,
    isPending: false
  };
}

/**
 * Helper used to return a canonical success result from mutation helpers.
 *
 * @returns `{ success: true }` — narrows {@link HookMutationResult} to {@link HookSuccessResult}.
 */
export function createHookSuccessResult() {
  return {
    success: true
  };
}

/**
 * Helper used to return a canonical failure result from mutation helpers.
 *
 * @param error - The {@link HookError} that caused the failure.
 * @returns `{ success: false, error }` — narrows to {@link HookFailureResult}.
 */
export function createHookFailureResult(error) {
  return {
    success: false,
    error
  };
}
