// react-native-sensitive-info/lib/module/hooks/useSecureStorage.js
"use strict";

import { useCallback, useMemo, useRef, useState } from 'react';
import { clearService, deleteItem, getAllItems, setItem } from "../core/storage.js";
import { normalizeStorageScopeOptions } from "../internal/options.js";
import deepEqual from "./internal/deepEqual.js";
import { createHookFailureResult, createHookSuccessResult } from "./types.js";
import useAsyncQuery from "./useAsyncQuery.js";
import useMutation from "./useMutation.js";

/**
 * Options accepted by {@link useSecureStorage}.
 */

const DEFAULTS = {
  includeValues: false,
  skip: false
};

/**
 * Removes hook-only flags so that mutation helpers receive pristine {@link SensitiveInfoOptions}.
 */
const stripIncludeValues = request => {
  const {
    includeValues: _includeValues,
    ...core
  } = request;
  return core;
};

// Frozen, shared empty list used as the items fallback so identity stays stable across renders.
const EMPTY_ITEMS = Object.freeze([]);

/**
 * Structure returned by {@link useSecureStorage}.
 */

/**
 * Manages a collection of secure items, exposing read/write helpers and render-ready state.
 *
 * Internally composes {@link useAsyncQuery} for the initial fetch and {@link useMutation} for
 * the imperative helpers, so the hook stays a thin choreography layer over the shared
 * lifecycle/abort/error machinery.
 *
 * @param options - Storage scoping plus hook-only flags (`includeValues`, `skip`).
 * @returns A {@link UseSecureStorageResult} with the cached `items`, lifecycle flags, and
 * imperative mutation helpers (`saveSecret`, `removeSecret`, `clearAll`, `refreshItems`).
 *
 * @remarks
 * - Mutations refresh the local cache automatically \u2014 you do **not** need to call `refreshItems`
 *   after `saveSecret`/`removeSecret`/`clearAll`.
 * - `clearAll` is **non-recoverable** \u2014 confirm intent in the UI before invoking.
 * - For single-key reads prefer {@link useSecret} or {@link useSecretItem}; this hook is optimized
 *   for managing a list of entries.
 *
 * @example
 * ```tsx
 * const { items, saveSecret, removeSecret, clearAll } = useSecureStorage({
 *   service: 'com.example.session',
 *   includeValues: true,
 * })
 *
 * await saveSecret('session-token', token)
 * ```
 *
 * @see {@link getAllItems}
 * @see {@link useSecret}
 */
export function useSecureStorage(options) {
  'use no memo';

  // Intentional opt-out: this hook coordinates several refs (cache, abort,
  // pending mutation) during render to keep the public API stable across
  // option-object identity changes — a pattern the React Compiler cannot
  // preserve without losing the deep-equality guarantees we ship.
  const fetchRunner = useCallback(request => {
    const includeValues = request.includeValues === true;
    return getAllItems(includeValues ? request : normalizeStorageScopeOptions(request));
  }, []);
  const fetchQuery = useAsyncQuery(fetchRunner, DEFAULTS, 'useSecureStorage.fetchItems', options, 'Ensure the service name matches the one used when storing the items.');
  const [localItems, setLocalItems] = useState(null);
  const items = useMemo(() => localItems ?? fetchQuery.data ?? EMPTY_ITEMS, [localItems, fetchQuery.data]);
  const {
    error: mutationError,
    mutate,
    clearError
  } = useMutation('useSecureStorage.mutate', '');

  // Stabilize the mutation options reference: callers commonly pass inline literals,
  // so we cache by structural equality to prevent cascading re-creation of saveSecret /
  // removeSecret / clearAll callbacks.
  const coreOptionsRef = useRef(null);
  const nextCoreOptions = stripIncludeValues({
    ...options
  });
  if (coreOptionsRef.current === null || !deepEqual(coreOptionsRef.current, nextCoreOptions)) {
    coreOptionsRef.current = nextCoreOptions;
  }
  const coreOptions = coreOptionsRef.current;
  const saveSecret = useCallback(async (key, value) => {
    const outcome = await mutate(() => setItem(key, value, coreOptions), {
      operation: 'useSecureStorage.saveSecret',
      hint: 'Check for duplicate keys or permission prompts that might have been dismissed.'
    });
    if (!outcome.success) {
      return createHookFailureResult(outcome.error);
    }
    setLocalItems(null);
    await fetchQuery.refetch();
    return createHookSuccessResult();
  }, [mutate, coreOptions, fetchQuery.refetch]);
  const removeSecret = useCallback(async key_0 => {
    const outcome_0 = await mutate(() => deleteItem(key_0, coreOptions), {
      operation: 'useSecureStorage.removeSecret',
      hint: 'Confirm the item still exists or that the user completed biometric prompts.'
    });
    if (!outcome_0.success) {
      return createHookFailureResult(outcome_0.error);
    }
    setLocalItems(prev => (prev ?? fetchQuery.data ?? []).filter(item => item.key !== key_0));
    return createHookSuccessResult();
  }, [mutate, coreOptions, fetchQuery.data]);
  const clearAll = useCallback(async () => {
    const outcome_1 = await mutate(() => clearService(coreOptions), {
      operation: 'useSecureStorage.clearAll',
      hint: 'Inspect whether another process holds a lock on the secure storage.'
    });
    if (!outcome_1.success) {
      return createHookFailureResult(outcome_1.error);
    }
    setLocalItems([]);
    clearError();
    return createHookSuccessResult();
  }, [mutate, coreOptions, clearError]);
  const refreshItems = useCallback(async () => {
    setLocalItems(null);
    await fetchQuery.refetch();
  }, [fetchQuery.refetch]);
  const error = mutationError ?? fetchQuery.error;
  return useMemo(() => ({
    items,
    isLoading: fetchQuery.isLoading,
    error,
    saveSecret,
    removeSecret,
    clearAll,
    refreshItems
  }), [items, fetchQuery.isLoading, error, saveSecret, removeSecret, clearAll, refreshItems]);
}
