// react-native-sensitive-info/lib/module/hooks/useSecurityAvailability.js
"use strict";

import { c as _c } from "react/compiler-runtime";
import { useCallback, useEffect, useMemo, useRef } from 'react';
import { AppState } from 'react-native';
import { getSupportedSecurityLevels } from "../core/storage.js";
import useAsync from "./useAsync.js";

/**
 * Tunables for {@link useSecurityAvailability}.
 *
 * @public
 */

/**
 * Result returned by {@link useSecurityAvailability}.
 *
 * @public
 */

const FOREGROUND_DEBOUNCE_MS = 500;

/**
 * Queries which security primitives are available on the current device and caches the outcome.
 *
 * @param options - Optional {@link UseSecurityAvailabilityOptions}; pass
 *   `{ refreshOnForeground: true }` to auto-refresh when the app returns to foreground.
 * @returns A {@link UseSecurityAvailabilityResult} with `data` (the latest snapshot),
 * `error`/`isLoading`/`isPending` flags, and a `refetch` helper that bypasses the cache.
 *
 * @remarks
 * - The hook caches the first successful response **per component instance** — subsequent
 *   renders of that same component reuse the cached value without hitting the native module.
 *   Multiple component instances each maintain their own cache.
 * - `refetch()` forces a fresh native call — use it after the user changes biometric enrollment
 *   in system settings.
 * - On error, the previously cached `data` is preserved so you can render fallback UI without
 *   losing capability info.
 * - When `refreshOnForeground` is enabled, the hook subscribes to `AppState` and refetches on
 *   `active` transitions (debounced). The subscription is created lazily inside `useEffect` so
 *   the hook remains tree-shakable and side-effect-free at the module level.
 *
 * @example
 * ```tsx
 * const { data: caps, isLoading } = useSecurityAvailability({ refreshOnForeground: true })
 *
 * if (isLoading || !caps) return null
 * if (caps.biometryStatus === 'notEnrolled') return <SetupFaceIdCta />
 * return caps.biometry
 *   ? <EnableFaceIdToggle />
 *   : <Text>Biometrics unavailable on this device.</Text>
 * ```
 *
 * @see {@link getSupportedSecurityLevels}
 * @see {@link useBiometryStatusWatcher}
 * @public
 */
export function useSecurityAvailability(options) {
  const $ = _c(14);
  const cacheRef = useRef(null);
  const forceRef = useRef(false);
  let t0;
  if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
    t0 = async () => {
      if (cacheRef.current && !forceRef.current) {
        return cacheRef.current;
      }
      forceRef.current = false;
      const value = await getSupportedSecurityLevels();
      cacheRef.current = value;
      return value;
    };
    $[0] = t0;
  } else {
    t0 = $[0];
  }
  const run = t0;
  let t1;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t1 = {
      hint: "Try calling SensitiveInfo.getSupportedSecurityLevels() directly to inspect the native error.",
      preserveDataOnError: true
    };
    $[1] = t1;
  } else {
    t1 = $[1];
  }
  const inner = useAsync(run, "useSecurityAvailability.fetch", t1);
  let t2;
  if ($[2] !== inner) {
    t2 = async () => {
      forceRef.current = true;
      await inner.refetch();
    };
    $[2] = inner;
    $[3] = t2;
  } else {
    t2 = $[3];
  }
  const refetch = t2;
  const refreshOnForeground = options?.refreshOnForeground === true;
  const lastRefreshRef = useRef(0);
  let t3;
  let t4;
  if ($[4] !== refetch || $[5] !== refreshOnForeground) {
    t3 = () => {
      if (!refreshOnForeground) {
        return;
      }
      const handleChange = next => {
        if (next !== "active") {
          return;
        }
        const now = Date.now();
        if (now - lastRefreshRef.current < FOREGROUND_DEBOUNCE_MS) {
          return;
        }
        lastRefreshRef.current = now;
        refetch();
      };
      const subscription = AppState.addEventListener("change", handleChange);
      return () => subscription.remove();
    };
    t4 = [refreshOnForeground, refetch];
    $[4] = refetch;
    $[5] = refreshOnForeground;
    $[6] = t3;
    $[7] = t4;
  } else {
    t3 = $[6];
    t4 = $[7];
  }
  useEffect(t3, t4);
  let t5;
  if ($[8] !== inner.data || $[9] !== inner.error || $[10] !== inner.isLoading || $[11] !== inner.isPending || $[12] !== refetch) {
    t5 = {
      data: inner.data,
      error: inner.error,
      isLoading: inner.isLoading,
      isPending: inner.isPending,
      refetch
    };
    $[8] = inner.data;
    $[9] = inner.error;
    $[10] = inner.isLoading;
    $[11] = inner.isPending;
    $[12] = refetch;
    $[13] = t5;
  } else {
    t5 = $[13];
  }
  return t5;
}
