// react-native-sensitive-info/lib/module/types.js
"use strict";

/**
 * @fileoverview React Native Sensitive Info v5.6.0 - Unified Type Definitions
 *
 * Comprehensive, strongly-typed interfaces optimized for:
 * - Tree-shaking: Zero unused code in production bundles
 * - VSCode autocomplete: Full JSDoc with examples
 * - Type safety: Discriminated unions prevent invalid combinations
 * - Performance: Minimal runtime overhead
 *
 * @version 5.6.0
 * @since 5.0.0
 */

/* ============================================================================
 * ACCESS CONTROL & AUTHENTICATION
 * ============================================================================ */

/**
 * Access control policy for secret access
 *
 * Determines which authentication method is required to access stored secrets.
 * The implementation automatically falls back to supported levels on the device.
 *
 * **Fallback Chain:**
 * `secureEnclaveBiometry` → `biometryCurrentSet` → `devicePasscode` → `software`
 *
 * @see {@link AuthenticationPrompt} for customizing auth dialogs
 *
 * @example
 * ```typescript
 * // Strongest: Secure Enclave + Biometric + Device Passcode
 * const policy: AccessControl = 'secureEnclaveBiometry';
 *
 * // Biometric only (revoke if biometrics change)
 * const currentSet: AccessControl = 'biometryCurrentSet';
 *
 * // Any enrolled biometric
 * const anyBio: AccessControl = 'biometryAny';
 *
 * // Device PIN/passcode required
 * const passcode: AccessControl = 'devicePasscode';
 *
 * // No additional protection (development only)
 * const none: AccessControl = 'none';
 * ```
 */

// No additional protection (weakest)

/**
 * Customizable text for biometric/passcode authentication prompts
 *
 * Shown when a user attempts to access a biometric-protected secret.
 * Localize these strings for your application's user base.
 *
 * **Security Note:** Don't reveal the specific secret being accessed!
 * Use generic descriptions like "Unlock your account" instead of
 * "Access payment token".
 *
 * @platform All platforms support these fields
 *
 * @example
 * ```typescript
 * // English
 * const prompt: AuthenticationPrompt = {
 *   title: 'Unlock Account',
 *   subtitle: 'Biometric authentication required',
 *   description: 'Verify your identity with Face ID or passcode',
 *   negativeButtonText: 'Cancel'
 * };
 *
 * // Spanish (localized)
 * const promptES: AuthenticationPrompt = {
 *   title: 'Desbloquear Cuenta',
 *   subtitle: 'Autenticación biométrica requerida',
 *   description: 'Verifica tu identidad con Face ID o contraseña'
 * };
 * ```
 */

/* ============================================================================
 * SECURITY METADATA & LEVELS
 * ============================================================================ */

/**
 * Actual security level achieved for stored secret
 *
 * Indicates the hardware backing and security guarantees.
 * **Ordered from strongest to weakest:**
 * 1. `secureEnclave` - iOS Secure Enclave (isolated processor)
 * 2. `strongBox` - Android StrongBox (dedicated secure processor)
 * 3. `biometry` - Biometric-protected keys
 * 4. `deviceCredential` - Device passcode/PIN protected
 * 5. `software` - Software-only encryption (weakest)
 *
 * **Decision Making:** Compare actual level with expected level.
 * If `actual < expected`, warn user or prompt for re-authentication.
 *
 * @example
 * ```typescript
 * const result = await setItem('token', value, {
 *   accessControl: 'secureEnclaveBiometry'
 * });
 *
 * if (result.metadata.securityLevel === 'software') {
 *   showWarning('⚠️ This device uses software encryption. ' +
 *     'Upgrade to iOS 16+ for hardware-backed security.');
 * }
 * ```
 */

// Software-only encryption (weakest)

/**
 * Metadata describing how a secret is stored and protected
 *
 * Provides visibility into the actual security achieved. Use this to:
 * - Validate security expectations
 * - Make app-level security decisions
 * - Provide user-facing security information
 * - Log security decisions for audit trails
 *
 * **Important:** The `securityLevel` field reflects what was **actually achieved**,
 * not what was **requested**. On older devices without Secure Enclave, requesting
 * `secureEnclaveBiometry` may result in `deviceCredential` being stored.
 *
 * @example
 * ```typescript
 * const { metadata } = await setItem('auth-token', token, {
 *   accessControl: 'secureEnclaveBiometry'
 * });
 *
 * // Check actual security level
 * const isStrongSecurity = metadata.securityLevel !== 'software';
 * const isHardwareBacked = ['secureEnclave', 'strongBox'].includes(
 *   metadata.securityLevel
 * );
 *
 * // Decide whether to accept this storage
 * if (!isStrongSecurity) {
 *   throw new Error('Security requirements not met');
 * }
 * ```
 */

/**
 * Device security capabilities available on current device
 *
 * Use to adapt UI/UX based on available security features.
 *
 * @example
 * ```typescript
 * const caps = await getSupportedSecurityLevels();
 *
 * if (caps.secureEnclave) {
 *   showSecureEnclaveOption(); // Show strongest option
 * }
 *
 * if (!caps.deviceCredential) {
 *   showSecurityWarning('Set device PIN for app protection');
 * }
 * ```
 */

/* ============================================================================
 * STORAGE OPTIONS & CONFIGURATION
 * ============================================================================ */

/**
 * Options for storing a secret
 *
 * Configure where, how, and with what protection a secret is stored.
 *
 * @example
 * ```typescript
 * const options: StorageOptions = {
 *   keychainService: 'myapp-auth',
 *   accessControl: 'secureEnclaveBiometry',
 *   authenticationPrompt: {
 *     title: 'Store Token',
 *     description: 'Authenticate to save your session'
 *   }
 * };
 * ```
 */

/**
 * Options for retrieving a stored secret
 *
 * Configure authentication prompt and service namespace.
 *
 * @example
 * ```typescript
 * const options: RetrievalOptions = {
 *   keychainService: 'myapp-auth',
 *   prompt: {
 *     title: 'Unlock Token',
 *     description: 'Authenticate to access your session'
 *   }
 * };
 * ```
 */

/* ============================================================================
 * OPERATION RESULTS
 * ============================================================================ */

/**
 * Result of a setItem operation
 *
 * Includes metadata about how the secret was stored.
 * **Always check metadata** to ensure security expectations were met!
 *
 * @example
 * ```typescript
 * const result = await setItem('token', value, {
 *   accessControl: 'secureEnclaveBiometry'
 * });
 *
 * console.log(`✓ Stored with ${result.metadata.securityLevel}`);
 *
 * if (result.metadata.securityLevel === 'software') {
 *   console.warn('⚠️ Not hardware-backed');
 * }
 * ```
 */

/**
 * A stored secret with its metadata
 *
 * Returned by getItem when successful.
 *
 * @example
 * ```typescript
 * const result = await getItem('auth-token', {
 *   keychainService: 'myapp'
 * });
 *
 * if (result) {
 *   console.log(`Token: ${result.value}`);
 *   console.log(`Stored: ${new Date(result.metadata.timestamp * 1000)}`);
 * }
 * ```
 */

/* ============================================================================
 * ERROR HANDLING
 * ============================================================================ */

/**
 * Error codes for SensitiveInfo operations
 *
 * All errors throw with a `code` property. Use to provide
 * appropriate user feedback and error recovery.
 *
 * **Categorized by severity:**
 * - User action: E_AUTH_CANCELED, E_AUTH_FAILED
 * - Key issue: E_KEY_INVALIDATED, E_NOT_FOUND
 * - System issue: E_KEYSTORE_UNAVAILABLE, E_ENCRYPTION_FAILED
 *
 * @example
 * ```typescript
 * try {
 *   const token = await getItem('auth-token', { ... });
 * } catch (error: any) {
 *   switch (error.code) {
 *     case 'E_AUTH_CANCELED':
 *       // User pressed cancel - normal flow
 *       console.log('User canceled');
 *       break;
 *
 *     case 'E_KEY_INVALIDATED':
 *       // User changed biometrics - ask to re-authenticate
 *       console.log('Please re-authenticate');
 *       break;
 *
 *     case 'E_KEYSTORE_UNAVAILABLE':
 *       // System error - show critical message
 *       showErrorAlert('Storage unavailable. Restart device.');
 *       break;
 *   }
 * }
 * ```
 */
export let ErrorCode = /*#__PURE__*/function (ErrorCode) {
  /**
   * The requested key doesn't exist
   *
   * This is a **normal condition**, not an error.
   * Recommended UX: Redirect to login/setup screen.
   */
  ErrorCode["NOT_FOUND"] = "E_NOT_FOUND";
  /**
   * User canceled biometric/passcode authentication
   *
   * This is expected user behavior. Don't treat as error.
   * Recommended UX: Close modal, allow retry.
   */
  ErrorCode["AUTH_CANCELED"] = "E_AUTH_CANCELED";
  /**
   * Biometric authentication failed (wrong fingerprint, etc)
   *
   * The OS allows limited retries before locking out.
   * Recommended UX: "Try again" or offer passcode fallback.
   */
  ErrorCode["AUTH_FAILED"] = "E_AUTH_FAILED";
  /**
   * Too many failed biometric attempts - locked out
   *
   * User must authenticate with device passcode.
   * Recommended UX: "Please use device passcode to unlock."
   */
  ErrorCode["BIOMETRY_LOCKOUT"] = "E_BIOMETRY_LOCKOUT";
  /**
   * The stored key has been invalidated
   *
   * Usually happens when user changes biometric enrollment.
   * iOS: Biometric enrollment changed
   * Android: Device lock settings changed
   * Recommended UX: Ask user to re-authenticate and re-save.
   */
  ErrorCode["KEY_INVALIDATED"] = "E_KEY_INVALIDATED";
  /**
   * Keystore/Keychain is unavailable or corrupted
   *
   * Rare - usually indicates serious system problems.
   * Recommended UX: Show critical error, suggest device restart.
   */
  ErrorCode["KEYSTORE_UNAVAILABLE"] = "E_KEYSTORE_UNAVAILABLE";
  /**
   * Encryption operation failed
   *
   * Possible causes:
   * - Invalid key
   * - Corrupted key material
   * - Out of memory
   * Recommended UX: Show error, offer factory reset option.
   */
  ErrorCode["ENCRYPTION_FAILED"] = "E_ENCRYPTION_FAILED";
  /**
   * Decryption operation failed
   *
   * Possible causes:
   * - Corrupted ciphertext
   * - Wrong key
   * - Data tampering detected
   * Recommended UX: Show error, offer to wipe and reset.
   */
  ErrorCode["DECRYPTION_FAILED"] = "E_DECRYPTION_FAILED";
  /**
   * Migration from v5.x format failed
   *
   * Rare - only occurs on first access if format migration fails.
   */
  ErrorCode["MIGRATION_FAILED"] = "E_MIGRATION_FAILED";
  return ErrorCode;
}({});

/**
 * Comprehensive error thrown by SensitiveInfo operations
 *
 * All SensitiveInfo errors include a `code` field for error categorization.
 *
 * @example
 * ```typescript
 * try {
 *   await getItem('key', { keychainService: 'myapp' });
 * } catch (error: any) {
 *   if (isSensitiveInfoError(error)) {
 *     console.error(`Error ${error.code}: ${error.message}`);
 *     console.error('Native error:', error.nativeError);
 *   }
 * }
 * ```
 */
export class SensitiveInfoError extends Error {
  /**
   * Machine-readable error code
   *
   * Use to make programmatic decisions.
   *
   * @see {@link ErrorCode} for list of all possible codes
   */

  /**
   * Original native error (iOS/Android)
   *
   * Useful for debugging. Contains platform-specific details.
   */

  /**
   * Creates a new SensitiveInfoError
   *
   * @param code - Machine-readable error code
   * @param message - Human-readable error message
   * @param nativeError - Optional original native error
   *
   * @example
   * ```typescript
   * const error = new SensitiveInfoError(
   *   ErrorCode.AUTH_FAILED,
   *   'Biometric authentication failed',
   *   originalError
   * );
   * throw error;
   * ```
   */
  constructor(code, message, nativeError) {
    super(message);
    this.name = 'SensitiveInfoError';
    this.code = code;
    this.nativeError = nativeError;

    // Maintain proper prototype chain for instanceof checks
    Object.setPrototypeOf(this, SensitiveInfoError.prototype);
  }
}

/**
 * Type guard for SensitiveInfoError
 *
 * Use to safely access error code in catch blocks.
 *
 * @example
 * ```typescript
 * try {
 *   await getItem('key', { ... });
 * } catch (error: unknown) {
 *   if (isSensitiveInfoError(error)) {
 *     // Now safe to access error.code
 *     console.log(error.code);
 *   }
 * }
 * ```
 */
export function isSensitiveInfoError(error) {
  return error instanceof SensitiveInfoError;
}

/* ============================================================================
 * CONSTANTS & DEFAULTS
 * ============================================================================ */

/**
 * Default access control level
 *
 * Use strongest available on device with automatic fallback.
 *
 * @default 'secureEnclaveBiometry'
 */
export const DEFAULT_ACCESS_CONTROL = 'secureEnclaveBiometry';

/**
 * Supported platforms
 *
 * @example
 * ```typescript
 * console.log(SUPPORTED_PLATFORMS);
 * // ["iOS", "macOS", "visionOS", "watchOS", "Android"]
 * ```
 */
export const SUPPORTED_PLATFORMS = ['iOS', 'macOS', 'visionOS', 'watchOS', 'Android'];

/**
 * Minimum supported OS versions per platform
 *
 * @example
 * ```typescript
 * console.log(MIN_VERSIONS.iOS); // "13.0"
 * console.log(MIN_VERSIONS.Android); // "8"
 * ```
 */
export const MIN_VERSIONS = {
  iOS: '13.0',
  macOS: '10.15',
  visionOS: '1.0',
  watchOS: '6.0',
  Android: '8'
};

/**
 * Maximum value length (in bytes)
 *
 * Exceeding this may cause storage failures or truncation.
 *
 * @default 10485760 (10 MB)
 */
export const MAX_VALUE_LENGTH = 10 * 1024 * 1024;

/**
 * Default keychain service name
 *
 * Used if service not specified in options.
 *
 * @default 'default'
 */
export const DEFAULT_KEYCHAIN_SERVICE = 'default';
