// react-native-sensitive-info/lib/commonjs/hooks/useAsync.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useAsync;
var _react = require("react");
var _errors = require("../errors.js");
var _errorUtils = _interopRequireDefault(require("./error-utils.js"));
var _types = require("./types.js");
var _useAsyncLifecycle = _interopRequireDefault(require("./useAsyncLifecycle.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const RESET_STATE = {
  data: null,
  error: null,
  isLoading: false,
  isPending: false
};
const reducer = (state, action) => {
  switch (action.type) {
    case 'reset':
      return state.data === null && state.error === null && !state.isLoading && !state.isPending ? state : RESET_STATE;
    case 'load':
      return state.isLoading && state.isPending ? state : {
        ...state,
        isLoading: true,
        isPending: true
      };
    case 'success':
      return {
        data: action.data,
        error: null,
        isLoading: false,
        isPending: false
      };
    case 'cancel':
      return state.error === null && !state.isLoading && !state.isPending ? state : {
        data: state.data,
        error: null,
        isLoading: false,
        isPending: false
      };
    case 'error':
      return {
        data: action.preserveData ? state.data : null,
        error: action.error,
        isLoading: false,
        isPending: false
      };
  }
};

/**
 * Consolidates the shared async-lifecycle pattern (state machine + abort + mount guard) used by
 * every data-fetching hook in this package.
 *
 * Callers must provide a **stable** `run` callback — typically wrapped in `useCallback` — along
 * with the operation identifier used for error reporting.
 *
 * @internal
 */
function useAsync(run, operation, options = {}) {
  'use no memo';

  // Intentional opt-out: the body uses optional chaining inside a try/catch
  // (a value-block pattern the React Compiler does not yet support). The
  // inner reducer + memoized callbacks already minimise re-renders.
  const {
    hint,
    skip = false,
    preserveDataOnError = false
  } = options;
  const [state, dispatch] = (0, _react.useReducer)(reducer, undefined, _types.createInitialAsyncState);
  const {
    begin,
    mountedRef
  } = (0, _useAsyncLifecycle.default)();
  const execute = (0, _react.useCallback)(async () => {
    if (skip) {
      dispatch({
        type: 'reset'
      });
      return;
    }
    const controller = begin();
    dispatch({
      type: 'load'
    });
    try {
      const result = await run(controller.signal);
      if (mountedRef.current && !controller.signal.aborted) {
        dispatch({
          type: 'success',
          data: result
        });
      }
    } catch (errorLike) {
      if (!mountedRef.current || controller.signal.aborted) return;
      if ((0, _errors.isAuthenticationCanceledError)(errorLike)) {
        dispatch({
          type: 'cancel'
        });
        return;
      }
      const hookError = (0, _errorUtils.default)(operation, errorLike, hint);
      dispatch({
        type: 'error',
        error: hookError,
        preserveData: preserveDataOnError
      });
    }
  }, [skip, begin, mountedRef, run, operation, hint, preserveDataOnError]);
  (0, _react.useEffect)(() => {
    execute().catch(() => {});
  }, [execute]);
  const refetch = (0, _react.useCallback)(async () => {
    await execute();
  }, [execute]);
  return (0, _react.useMemo)(() => ({
    data: state.data,
    error: state.error,
    isLoading: state.isLoading,
    isPending: state.isPending,
    refetch
  }), [state.data, state.error, state.isLoading, state.isPending, refetch]);
}
